Feature/triplestore wrapper

dsmits opened the pull request:
Created custom wrapper that queries a sparql endpoint rather than csv file.

It will convert the data into a pandas dataframe and hand it to the vantage6 algorithm in the same way that the conventional wrapper does

dsmits committed:
Made initial implementation stub with test

dsmits committed:
Created dockerfile for sparql wrapper
Fixed wrapper after testing in infrastructure
Added SPARQLwrapper to requirements

dsmits committed:
Fixed unittest

svenvanderburg commented on Dockerfile.sparql:
```diff
@@ -0,0 +1,15 @@
+# basic python3 image as base
```
Doesn't seem like a basic python3 image, but a vantage6 one?

svenvanderburg commented on Dockerfile.sparql:
```diff
@@ -0,0 +1,15 @@
+
+ENV PKG_NAME=${PKG_NAME}
+
+# Tell docker to execute `docker_wrapper()` when the image is run.
```
This runs sparql_wrapper, not docker_wrapper?

svenvanderburg commented on v6_carrier_py/wrapper.py:
```diff
@@ -0,0 +1,87 @@
+from vantage6.client import serialization
+from vantage6.tools import docker_wrapper
+from vantage6.tools.data_format import DataFormat
+from vantage6.tools.dispatch_rpc import dispact_rpc
```
Is this really a typo in the vantage6 library?

dsmits replied:
yeah it really is, I know :disappointed:

svenvanderburg commented on v6_carrier_py/wrapper.py:
```diff
@@ -0,0 +1,87 @@
+
+    endpoint = os.environ["DATABASE_URI"]
+
+    # Workaround because the endpoint is automatically prefixed with the data folder. However this does not make
+    # sense for a sparql endpoint.
```
I would prefer having this documentation in `_fix_endpoint`

svenvanderburg commented on v6_carrier_py/wrapper.py:
```diff
@@ -0,0 +1,87 @@
+
+    # write output from the method to mounted output file. Which will be
+    # transfered back to the server by the node-instance.
+    output_file = os.environ["OUTPUT_FILE"]
```
Maybe you can factor this out into a function: write_output?

dsmits replied:
This code is actually a copy of code I wrote for the regular vantage6 wrapper. Ideally I should make that into a function, along with some other reusable stuff. I'd say that belongs in a different PR because I first have to create a PR for fixing this in the regular vantage6 wrapper

svenvanderburg replied:
Agree!

svenvanderburg commented on v6_carrier_py/wrapper.py:
```diff
@@ -0,0 +1,87 @@
+SPARQL_RETURN_FORMAT = CSV
+
+
+def sparql_wrapper(module: str):
```
Can you add some documentation here? What is module?

svenvanderburg replied:
It would also be good to document what the query should look like.

dsmits replied:
I specified that it should be a SPARQL `SELECT` query string. That seems enough to me.

svenvanderburg replied:
Nice, I really like the documentation! It would have helped me if you specified what the query should return a bit more, but I'll address that in a different comment.

svenvanderburg commented on v6_carrier_py/wrapper.py:
```diff
@@ -0,0 +1,87 @@
+
+    info(f"Using '{endpoint}' as triplestore endpoint")
+
+    data = query_triplestore(endpoint, query)
```
You will most likely end up with a hard-to-interpret list of triples that you then need to convert to a usable format, I am not sure if a pandas dataframe is the best way to send it to the algorithm. The data is graph data and not spreadsheet data. So maybe pass a rdflib.Graph() object instead of a pd.DataFrame?

If you are doing the pandas dataframe so you can run the same algorithms with different wrappers, isn't that a bit wishful thinking? I.e. the fact that it is in a pandas dataframe doesn't mean it is meaningful spreadsheet data.

dsmits replied:
I think you are misunderstanding the output of the query. We are querying graph data, yes, but the output doesn't have to be triples, a bit similar to how a sql query does not have to return all the rows in a table, just the ones you selected. 

Example query (simplified version from an [example from wikidata](https://www.wikidata.org/wiki/Wikidata:SPARQL_query_service/queries/examples#Goats), try it out if you like):

```
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
SELECT ?item ?itemLabel 
WHERE 
{
  # Item should be instance of goat (wd:Q2934)
  ?item wdt:P31 wd:Q2934.
  ?item  skos:prefLabel ?itemLabel
}
```

svenvanderburg replied:
Yes, you are completely right. I didn't think about in that way (my knowledge is very limited when it comes to SPARQL queries). After your demo this was more clear, and I wrote [this comment](https://github.com/CARRIER-project/vantage6-algorithms/pull/20#issuecomment-717331325).

I think the crux is that this way of thinking about RDF triples is now quite natural for you, but it clearly isn't for me, and might also not be for data scientists just wanting to run an algorithm. So I would recommend documenting the kind of queries that you expect, and provide this example.

svenvanderburg replied:
Nice, I like the example! :)

svenvanderburg reviewed (commented):
Nice work!

I have added some minor comments to the file changes.

Some more thoughts:
- I think this is a good addition. I still think that it is quite difficult to build an algorithm on top of this compared to CSV. You will always need to do some more custom processing to get the graph data in a usable format. It would be interesting to see how we can make this easier for an algorithm developer, but I guess we cannot really.
- This is a workaround for having direct access to the SPARQL endpoint right? Because that is not possible from the algorithm container? Otherwise I would actually prefer that over using a wrapper like this.

svenvanderburg commented:
Your demo made me think: could you actually write a SPARQL query that gives back meaningful tabular data (instead of the s, p, o example that you showed)? In that case you could restrict using this wrapper with those kind of queries maybe (and then do the conversion to pandas)?

dsmits commented:
> Nice work!
> 
> I have added some minor comments to the file changes.
> 
> Some more thoughts:
> 
>     * I think this is a good addition. I still think that it is quite difficult to build an algorithm on top of this compared to CSV. You will always need to do some more custom processing to get the graph data in a usable format. It would be interesting to see how we can make this easier for an algorithm developer, but I guess we cannot really.
> 
>     * This is a workaround for having direct access to the SPARQL endpoint right? Because that is not possible from the algorithm container? Otherwise I would actually prefer that over using a wrapper like this.

To address you points:
1. I don't see how it is more difficult to build an algorithm on top of this than CSV. The idea is that the user writes the SPARQL query. The algorithm will then be run on the output of that query. 
RDF literals [can have numerical datatypes as well.](https://www.w3.org/2011/rdf-wg/wiki/XSD_Datatypes) .
The algorithms just get a pandas dataframe, same as with the csv approach.

2. The wrapper is actually running within the algorithm container. If you would write a raw custom algorithm without a wrapper you could also access the sparql endpoint. The benefit of this wrapper is that it separates the algorithms from the data retrieval. No matter which wrapper you use, the csv one or the sparql one, you only have to implement the algorithms once. The new wrapper is compatible with all the old algorithms that I implemented for the csv case.

dsmits committed:
Added count algorithm to count rows

dsmits committed:
Improved documentation

svenvanderburg commented:
OK, I totally see the benefits now :) Sorry for being so critical! Indeed it is really nice that you can separate the algorithms from data retrieval.

dsmits committed:
More explanation

svenvanderburg reviewed (approved):
The documentation makes it really clear now what the rationale is of using this and how it should be used. So good job!
